crates/revm/src/executor.rs

```diff
@@ -227,46 +230,175 @@ where
 
         self.init_env(&block.header, total_difficulty);
 
+        #[cfg(feature = "optimism")]
+        let mut l1_block_info = optimism::L1BlockInfo::try_from(block)?;
+
         let mut cumulative_gas_used = 0;
         let mut post_state = PostState::with_tx_capacity(block.number, block.body.len());
         for (transaction, sender) in block.body.iter().zip(senders) {
-            // The sum of the transaction’s gas limit, Tg, and the gas utilised in this block prior,
-            // must be no greater than the block’s gasLimit.
-            let block_available_gas = block.header.gas_limit - cumulative_gas_used;
-            if transaction.gas_limit() > block_available_gas {
-                return Err(BlockValidationError::TransactionGasLimitMoreThanAvailableBlockGas {
-                    transaction_gas_limit: transaction.gas_limit(),
-                    block_available_gas,
+            #[cfg(feature = "optimism")]
+            {
+                let db = self.db();
```

Comment:
@rakita @gakonst do you think we can support these behind a config value in revm?

Reply:
An alternative: we could move the changes here to a different function "execute_transactions_optimism()" to keep this function cleaner. Would make the feature flag less annoying but add some duplicated code. Lmk

Reply:
To add to this: me and @clabby initially thought of adding the changes to revm directly, however after looking at it, there are a few reasons why doing it in Reth could be better.

Referencing https://github.com/refcell/reth/pull/6#discussion_r1240710989

> This could have been addressed in Revm's transact() function in a more clean way. In fact I have a local branch of what this change would look like here: https://github.com/merklefruit/revm/pull/1/files#diff-1d478ba44ccc56e3b1142bd3723bf97f3e254c25dd18323481aedadce0803e91R130

> However that solution has some cons:

> 1. we are not able to revert the deposit if the block runs out of gas, because revm's transact() context doesn't know about the outstanding block gas used (called cumulative_gas_used here)

> 2. we would need to reimplement the logic to extract l1_cost from the first tx of the block, which we have to do in reth anyway (both here and in receipts)

Reply:
^ Yep, the initial reason we decided to put these changes in reth rather then revm are covered above. Keeping the diff in reth feels a bit cleaner as well, but there is a good bit of code duplication. I'm a fan of moving this logic to a different function + pulling shared logic into generic funcs.

Reply:
btw there is a bigger refactor taking place currently that introduces a different way of how the state is handled. There will be a lot of conflicts here: https://github.com/paradigmxyz/reth/pull/3512
And associated revm change with new state: https://github.com/bluealloy/revm/pull/499

Reply:
Resolving this conversation to clean up this pr. Added a note to check these change in the top-level pr description under reviewer notes.